# Data Science with Data Frames
In this notebook, we will be using the powerful tool of dataframes (a pandas object) to investigate data from two sources: Taylor Swift's discography and a database of world crime stats. We will focusing on descriptive statistics, which are statistics that describe the dataset, like mean and max.

In [46]:
# import the NumPy library
import numpy as np
# import the Pandas library
import pandas as pd

## Taylor Swift
We will begin this chapter by analyzing a dataset from Spotify, which contains information about all of the tracks uploaded by Taylor Swift's Spotify account. The first step is to load in our data, which we can do using the read_csv command.

In [5]:
# read the dataset into the DataFrame
df = pd.read_csv('taylor_swift_spotify_data.csv')

Now that we have the data read into our notebook, let's look at the data itself. We can do this by printing the column names, or even just checking the first few values of the data table.

In [6]:
# access the columns
columns = df.columns
print(columns)

Index(['artist_name', 'artist_id', 'album_id', 'album_type',
       'album_release_date', 'album_release_year',
       'album_release_date_precision', 'danceability', 'energy', 'key',
       'loudness', 'mode', 'speechiness', 'acousticness', 'instrumentalness',
       'liveness', 'valence', 'tempo', 'track_id', 'analysis_url',
       'time_signature', 'disc_number', 'duration_ms', 'explicit',
       'track_href', 'is_local', 'track_name', 'track_preview_url',
       'track_number', 'type', 'track_uri', 'external_urls.spotify',
       'album_name', 'key_name', 'mode_name', 'key_mode'],
      dtype='object')


In [7]:
df.head()

artist_name               artist_id                album_id album_type  \
0  Taylor Swift  06HL4z0CvFAxyc27GXpf02  3lS1y25WAhcqJDATJK70Mq      album   
1  Taylor Swift  06HL4z0CvFAxyc27GXpf02  3lS1y25WAhcqJDATJK70Mq      album   
2  Taylor Swift  06HL4z0CvFAxyc27GXpf02  3lS1y25WAhcqJDATJK70Mq      album   
3  Taylor Swift  06HL4z0CvFAxyc27GXpf02  3lS1y25WAhcqJDATJK70Mq      album   
4  Taylor Swift  06HL4z0CvFAxyc27GXpf02  3lS1y25WAhcqJDATJK70Mq      album   

  album_release_date  album_release_year album_release_date_precision  \
0         2022-10-22                2022                          day   
1         2022-10-22                2022                          day   
2         2022-10-22                2022                          day   
3         2022-10-22                2022                          day   
4         2022-10-22                2022                          day   

   danceability  energy  key  ...                              track_name  \
0         0.735   0.444   10  ...                           Lavender Haze   
1         0.658   0.378    7  ...                                  Maroon   
2         0.638   0.634    4  ...                               Anti-Hero   
3         0.659   0.323    9  ...  Snow On The Beach (feat. Lana Del Rey)   
4         0.694   0.380    2  ...                 You're On Your Own, Kid   

   track_preview_url  track_number   type  \
0                NaN             1  track   
1                NaN             2  track   
2                NaN             3  track   
3                NaN             4  track   
4                NaN             5  track   

                              track_uri  \
0  spotify:track:4g2c7NoTWAOSYDy44l9nub   
1  spotify:track:199E1RRrVmVTQqBXih5qRC   
2  spotify:track:02Zkkf2zMkwRGQjZ7T4p8f   
3  spotify:track:6ADDIJxxqzM9LMpm78yzQG   
4  spotify:track:7gVWKBcfIW93YxNBi3ApIE   

                               external_urls.spotify               album_name  \
0  https://open.spotify.com/track/4g2c7NoTWAOSYDy...  Midnights (3am Edition)   
1  https://open.spotify.com/track/199E1RRrVmVTQqB...  Midnights (3am Edition)   
2  https://open.spotify.com/track/02Zkkf2zMkwRGQj...  Midnights (3am Edition)   
3  https://open.spotify.com/track/6ADDIJxxqzM9LMp...  Midnights (3am Edition)   
4  https://open.spotify.com/track/7gVWKBcfIW93YxN...  Midnights (3am Edition)   

   key_name mode_name  key_mode  
0        A#     major  A# major  
1         G     major   G major  
2         E     major   E major  
3         A     major   A major  
4         D     major   D major  

[5 rows x 36 columns]

how many columns are in this data table? how many rows (i.e., data points)?

In [5]:
# how many columns are there, how many rows?
print('No. of columns:', len(columns))
print('No. of rows:', len(df))

No. of columns: 36
No. of rows: 1265


now, let's move on to generating some descriptive statistics (statistical quantities that describe the dataset)

In [8]:
# what's the average length of a Taylor Swift song?
avg_duration = df['duration_ms'].mean() # we can use the mean function for this
print(avg_duration)
# this is great, but it's in milliseconds. who knows about milliseconds?

230381.6181818182


In [7]:
# define a function to convert the millisecond timestamp into something more understandable
def convert_time(time_ms):
    # convert from milliseconds to seconds
    time_sec = time_ms / 1000
    # get the minutes, dividing the seconds by 60
    time_min = int(time_sec / 60)
    # get the seconds extra past the last minute
    time_sec = int(time_sec % 60)
    return str(time_min)+' minutes and '+str(time_sec)+' seconds'

print('The average song length is ', convert_time(avg_duration))

The average song length is  3 minutes and 50 seconds


Often, the purpose of generating descriptive statistics is to see how our dataset compares to another. So, how does our result for the average Taylor Swift song length compare to other modern pop songs? Well, the average in recent years is about... 3 minutes and 50 seconds, too!

https://www.vox.com/2014/8/18/6003271/why-are-songs-3-minutes-long

How about the tempo, then? the average tempo is 116 bpm

https://www.washingtonpost.com/news/to-your-health/wp/2015/10/30/the-mathematical-formula-behind-feel-good-songs/

In [8]:
df['tempo'].mean()
# a bit faster, but still very typical of the pop genre

120.87697944664023

what are the top 3 keys that Taylor Swift likes to write in?

In [9]:
df['key_mode'].value_counts()[:3]

G major    224
F major    156
C major    151
Name: key_mode, dtype: int64

### Data Cleaning
Often, the original / raw dataset that we get from the real world is not exactly what we want to analyze. For example, let's list all of Taylor Swift's albums on Spotify

In [12]:
df['album_name'].unique()

array(['Midnights', 'evermore', 'folklore', 'Lover',
       'Taylor Swift Karaoke: reputation', 'reputation',
       'reputation Stadium Tour Surprise Song Playlist', '1989',
       'Taylor Swift Karaoke: 1989', 'Red', 'Taylor Swift Karaoke: Red',
       'Speak Now World Tour Live', 'Speak Now',
       'Taylor Swift Karaoke: Speak Now', 'Fearless', 'Fearless Karaoke',
       'Fearless Platinum Edition',
       'Live From Clear Channel Stripped 2008', 'Taylor Swift',
       'Taylor Swift Karaoke'], dtype=object)

it seems like there are lots of albums that have several versions. can we clean out this data?


In [10]:
# create a mask that gets rid of any () versions
mask = ~df['album_name'].str.contains('\(')
albums = df[mask]['album_name'].unique()
print(albums) # great! That's much better...

['Midnights' 'evermore' 'folklore' 'Lover'
 'Taylor Swift Karaoke: reputation' 'reputation'
 'reputation Stadium Tour Surprise Song Playlist' '1989'
 'Taylor Swift Karaoke: 1989' 'Red' 'Taylor Swift Karaoke: Red'
 'Speak Now World Tour Live' 'Speak Now' 'Taylor Swift Karaoke: Speak Now'
 'Fearless' 'Fearless Karaoke' 'Fearless Platinum Edition'
 'Live From Clear Channel Stripped 2008' 'Taylor Swift'
 'Taylor Swift Karaoke']


I am trying to open a club and I need some music to play. What's Taylor Swift's most danceable song?

In [12]:
# find the maximum of danceability metric, which is another descriptive statistic
max_dance = df['danceability'].max()
# generate a mask to filter the data
mask = df['danceability'] == max_dance
# now, display the masked data (but only the three relevant columns)
df[mask][['album_name', 'track_name', 'danceability']]

album_name        track_name  danceability
370      Lover  I Think He Knows         0.897

great job with this! now, we're going to move on to another example

## World Crime Index
The world crime index is a way of measuring how dangerous cities are, using a unitless index out of 100. The higher the crime index (and thus the lower the safety index is), the more dangerous the city. Let's start, just like we did with Taylor Swift, by reading in the data.

In [75]:
df = pd.read_csv('world_crime_index.csv')
df

Rank                             City  Crime Index  Safety Index
0       1               Caracas, Venezuela        83.98         16.02
1       2           Pretoria, South Africa        81.98         18.02
2       3                   Celaya, Mexico        81.80         18.20
3       4         San Pedro Sula, Honduras        80.87         19.13
4       5   Port Moresby, Papua New Guinea        80.71         19.29
..    ...                              ...          ...           ...
448   449              Quebec City, Canada        15.14         84.86
449   450                   Taipei, Taiwan        15.05         84.95
450   451             San Sebastian, Spain        14.86         85.14
451   452                      Doha, Qatar        13.96         86.04
452   453  Abu Dhabi, United Arab Emirates        11.67         88.33

[453 rows x 4 columns]

It looks like we have 453 cities in our database, and it's already sorted by most dangerous (descending). Can we find out which country has the safest cities? We'll have to make a new column in our data frame, since "City" includes both the city and the country.

In [76]:
# let's start by looking at one City entry. Can you think of a way to get the country?
df['City'][0]

'Caracas, Venezuela'

In [77]:
# add in the country data
df['Country'] = [x.split(', ')[1] for x in df['City']]
# now, list all of the countries found in the dataset
countries = df['Country'].unique()
countries

array(['Venezuela', 'South Africa', 'Mexico', 'Honduras',
       'Papua New Guinea', 'Afghanistan', 'Brazil', 'Trinidad And Tobago',
       'MD', 'Argentina', 'TN', 'MI', 'Australia', 'Jamaica',
       'United Kingdom', 'NM', 'Peru', 'Ecuador', 'MO', 'El Salvador',
       'Colombia', 'Namibia', 'Puerto Rico', 'Dominican Republic',
       'Syria', 'Angola', 'LA', 'WI', 'CA', 'Nigeria', 'IL', 'France',
       'Philippines', 'Canada', 'OH', 'TX', 'Kazakhstan', 'Maldives',
       'AB', 'Bangladesh', 'PA', 'Italy', 'Malaysia', 'GA', 'Guatemala',
       'Libya', 'AK', 'Zimbabwe', 'Romania', 'Tanzania', 'India', 'Chile',
       'DC', 'Iraq', 'Kenya', 'Belarus', 'IN', 'Belgium', 'FL', 'Sweden',
       'Greece', 'KY', 'Iran', 'Botswana', 'Morocco', 'Costa Rica',
       'Mongolia', 'MN', 'WA', 'NV', 'Uruguay', 'Spain', 'Pakistan',
       'Algeria', 'Vietnam', 'Cambodia', 'Indonesia', 'Ukraine',
       'Portugal', 'Ireland', 'OR', 'Russia', 'AZ', 'Egypt', 'Paraguay',
       'BC', 'NY', 'VA', 'Tur

uh oh! it seems like our method split the USA into states, which we did not want. how can we consider the whole USA all at once?

In [78]:
# add in the country data, this time indexing by -1
df['Country'] = [x.split(', ')[-1] for x in df['City']]
# now, list all of the countries found in the dataset
countries = df['Country'].unique()
countries

array(['Venezuela', 'South Africa', 'Mexico', 'Honduras',
       'Papua New Guinea', 'Afghanistan', 'Brazil', 'Trinidad And Tobago',
       'United States', 'Argentina', 'Australia', 'Jamaica',
       'United Kingdom', 'Peru', 'Ecuador', 'El Salvador', 'Colombia',
       'Namibia', 'Puerto Rico', 'Dominican Republic', 'Syria', 'Angola',
       'Nigeria', 'France', 'Philippines', 'Canada', 'Kazakhstan',
       'Maldives', 'Bangladesh', 'Italy', 'Malaysia', 'Guatemala',
       'Libya', 'Zimbabwe', 'Romania', 'Tanzania', 'India', 'Chile',
       'Iraq', 'Kenya', 'Belarus', 'Belgium', 'Sweden', 'Greece', 'Iran',
       'Botswana', 'Morocco', 'Costa Rica', 'Mongolia', 'Uruguay',
       'Spain', 'Pakistan', 'Algeria', 'Vietnam', 'Cambodia', 'Indonesia',
       'Ukraine', 'Portugal', 'Ireland', 'Russia', 'Egypt', 'Paraguay',
       'Turkey', 'Montenegro', 'Panama', 'Ethiopia', 'Tunisia', 'Ghana',
       'North Macedonia', 'New Zealand', 'Lebanon', 'Thailand',
       'Bosnia And Herzegovina', 

In [79]:
# first, we'll create an array full of zeros to store this data. what size should it be?
country_avg_safety = np.zeros(len(countries))
for i in range(0, len(countries)):
    mask = df['Country'] == countries[i]
    country_avg_safety[i] = df[mask]['Safety Index'].mean()

In [80]:
df = pd.DataFrame({'Country': countries, 'Average Safety': country_avg_safety})
# based on the cities in this dataset, which country is the safest?
df.sort_values('Average Safety', ascending = False).head(1)

Country  Average Safety
117   Qatar           86.04

based on this data, it seems like Qatar has the safest streets!

limitations:
* criminal statutes differ across countries
* not all cities included
* soft crime like harrassment would not be considered

note: it's always important to consider the limitations of your dataset as a data scientist